src/memory_region.rs

```diff
@@ -52,68 +52,102 @@ impl fmt::Debug for MemoryRegion {
     }
 }
 
-/// Helper for translate_addr to generate errors
-fn generate_access_violation<E: UserDefinedError>(
-    vm_addr: u64,
-    len: usize,
-    access_type: &str,
-    pc: usize,
-    regions: &[MemoryRegion],
-) -> EbpfError<E> {
-    let mut regions_string = "".to_string();
-    if !regions.is_empty() {
-        regions_string = "regions:".to_string();
-        for region in regions.iter() {
-            regions_string = format!(
-                "  {} \n{:#x}-{:#x}",
-                regions_string,
-                region.addr_vm,
-                region.addr_vm + region.len - 1,
-            );
+/// Type of memory access
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum AccessType {
+    /// Read
+    Load,
+    /// Write
+    Store,
+}
+
+/// Indirection to use instead of a slice to make handling easier
+#[derive(Default)]
+pub struct MemoryMapping {
```

Comment:
If we know the final size of `MemoryMapping`'s vector lets use `with_capacity` to reduce allocations